helix-gateway/src/main/java/org/apache/helix/gateway/grpcservice/HelixGatewayServiceService.java

```diff
@@ -16,6 +27,13 @@ public StreamObserver<proto.org.apache.helix.gateway.HelixGatewayServiceOuterCla
       public void onNext(ShardStateMessage request) {
         // called when a client sends a message
         //....
+        String instanceName = request.getInstanceName();
+        if (_observerMap.containsValue(instanceName)) {
+          // this is from an existing client
+        } else {
+          updateObserver(instanceName, responseObserver);
+          // update state map
+        }
```

Comment:
Let's do this:

if (! contains) {
   update
}

do regular thing.

Reply:
TFTR updated.